Lombiq.Tests.UI.Samples/Tests/BasicVisualTests.cs

```diff
@@ -0,0 +1,173 @@
+                // Just like above, save and append it to failure dump.
+                SaveImageToTempAndAppendToFailureDump(
+                    referenceImage,
+                    context,
+                    Path.Combine(testTempRootFolder, referenceImageCroppedFileName),
+                    Path.Combine(testDumpFolderName, referenceImageCroppedFileName));
+
+                // At this point, we have reference and captured images too.
+                // Creating diff image is not required, but it can be very useful to investigate failing tests.
+                // You can read more about how diff created here:
+                // https://github.com/Codeuctivity/ImageSharp.Compare/blob/1.2.11/ImageSharpCompare/ImageSharpCompare.cs#L258.
+                // So lets create it and append it to failure dump.
+                using var diffImage = referenceImage
+                    .CalcDiffImage(navbarImage)
+                    .ShouldNotBeNull();
+                var diffImageFileName = "navbar_diff.bmp";
+
+                SaveImageToTempAndAppendToFailureDump(
+                    diffImage,
+                    context,
+                    Path.Combine(testTempRootFolder, diffImageFileName),
+                    Path.Combine(testDumpFolderName, diffImageFileName));
+
+                // Now we are one step far from the end. Here we create a statistical summary of differences between the
+                // captured and the reference image. At the end, the lower values are better.
+                // You can read more about how this statistical calculations are created here:
+                // https://github.com/Codeuctivity/ImageSharp.Compare/blob/1.2.11/ImageSharpCompare/ImageSharpCompare.cs#L143.
+                var diff = referenceImage
+                    .CompareTo(navbarImage);
+                var diffLogFileName = "navbar_diff.log";
+                var diffLogTempFileName = Path.Combine(testTempRootFolder, diffLogFileName);
+
+                File.WriteAllText(
+                    diffLogTempFileName,
+                    string.Format(
+                        CultureInfo.InvariantCulture,
+                        @"
+navbar: calculated differences:
+    absoluteError={0},
+    meanError={1},
+    pixelErrorCount={2},
+    pixelErrorPercentage={3}",
+                        diff.AbsoluteError,
+                        diff.MeanError,
+                        diff.PixelErrorCount,
+                        diff.PixelErrorPercentage));
+                context.AppendFailureDump(
+                    Path.Combine(testDumpFolderName, diffLogFileName),
+                    context => Task.FromResult((Stream)File.OpenRead(diffLogTempFileName)));
+
+                // All the stuff above are made for this comparison. Here we check that, the erroneous pixels percentage
+                // is less than the given threshold.
+                diff.PixelErrorPercentage.ShouldBeLessThan(8);
```

Comment:
Let's move all the above complexity behind a simple API, similar to [Shouldly's `MatchApproved()`](https://docs.shouldly.io/documentation/equality/matchapproved), so we don't have to do and think through all this every time (the two occasions are already too complex).

I'd imagine something like `context.AssertVisualVerification()`. This would take optional parameters for selecting an element and tolerance (i.e. this `ShouldBeLessThan()` check). This is how you'd use it:

1. Let's suppose you run the test locally the first time. It fails, because there's no approved file, but it creates one under the test project with a conventional name, that's unique for the given assertion (again check `MatchApproved()`, it does exactly this). You get a helpful error message that the failure is due to the approved file not existing, but all subsequent runs will match against this file and pass if there's no diff too large.
2. You run it again. This time, the test passes, since the approved file exists and matches.
3. You change something in the app and run the test again, making it fail due to a diff too large. The test fails with a descriptive error message that explains what failed, and where you can find more information, what's in the dump. It also tells that if you want to override the existing accepted file, just delete it and run the test again.

What do you think?